File: packages/shared/src/notification.ts
/**
 * Shared notification types for Jarvis mobile and desktop apps.
 *
 * These types are used by both the backend (for sending notifications)
 * and the frontend/mobile apps (for receiving and displaying them).
 */

/** Supported push notification event types */
export type NotificationEventType =
  | 'test_run_completed'
  | 'knowledge_digest_ready'
  | 'freelance_task_assigned'
  | 'new_message';

/** Device platform types */
export type DevicePlatform = 'ios' | 'android' | 'web' | 'desktop';

/** A push notification payload sent from the backend */
export interface NotificationPayload {
  event_type: NotificationEventType;
  title: string;
  body: string;
  data?: Record<string, string>;
  sent_at?: string;
}

/** A registered device token */
export interface DeviceToken {
  id: string;
  token: string;
  platform: DevicePlatform;
  device_name?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

/** User notification preferences */
export interface NotificationPreferences {
  test_run_completed: boolean;
  knowledge_digest_ready: boolean;
  freelance_task_assigned: boolean;
  new_message: boolean;
}

/** API request to register a device token */
export interface RegisterDeviceRequest {
  token: string;
  platform: DevicePlatform;
  device_name?: string;
}

/** API request to update notification preferences */
export interface UpdatePreferencesRequest {
  test_run_completed?: boolean;
  knowledge_digest_ready?: boolean;
  freelance_task_assigned?: boolean;
  new_message?: boolean;
}

/** Notification event type display labels */
export const NOTIFICATION_LABELS: Record<NotificationEventType, string> = {
  test_run_completed: 'Test Run Completed',
  knowledge_digest_ready: 'Knowledge Digest Ready',
  freelance_task_assigned: 'Freelance Task Assigned',
  new_message: 'New Message',
};

/** Notification event type icons (emoji) */
export const NOTIFICATION_ICONS: Record<NotificationEventType, string> = {
  test_run_completed: '✅',
  knowledge_digest_ready: '📚',
  freelance_task_assigned: '📋',
  new_message: '💬',
};

File: backend/app/schemas/notification.py
"""Pydantic schemas for push notifications."""

from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel, Field


class DeviceTokenRegisterRequest(BaseModel):
    """Request to register a device token for push notifications."""

    token: str = Field(..., min_length=1, max_length=500)
    platform: str = Field("unknown", pattern=r"^(ios|android|web|desktop|unknown)$")
    device_name: str | None = Field(None, max_length=200)


class DeviceTokenResponse(BaseModel):
    """Response schema for a device token."""

    id: UUID
    token: str
    platform: str
    device_name: str | None = None
    is_active: bool = True
    created_at: str
    updated_at: str

    model_config = {"from_attributes": True}


class DeviceTokenListResponse(BaseModel):
    """List of registered device tokens."""

    devices: list[DeviceTokenResponse]
    total: int


class NotificationPreferencesResponse(BaseModel):
    """Response schema for notification preferences."""

    test_run_completed: bool = True
    knowledge_digest_ready: bool = True
    freelance_task_assigned: bool = True
    new_message: bool = True

    model_config = {"from_attributes": True}


class NotificationPreferencesUpdateRequest(BaseModel):
    """Request to update notification preferences."""

    test_run_completed: bool | None = None
    knowledge_digest_ready: bool | None = None
    freelance_task_assigned: bool | None = None
    new_message: bool | None = None


class SendNotificationRequest(BaseModel):
    """Request to send a notification (for internal use)."""

    user_id: str
    event_type: str = Field(..., pattern=r"^(test_run_completed|knowledge_digest_ready|freelance_task_assigned|new_message)$")
    title: str = Field(..., min_length=1, max_length=200)
    body: str = Field(..., min_length=1)
    data: dict | None = None